### ISSUE TITLE
TemporalFusionTransformer non deterministic training

### ISSUE THREAD

**Issue by BorisAnc** [jdb78/pytorch-forecasting#219]
- PyTorch-Forecasting version: 0.7.1
- PyTorch version: 1.7.1
- Python version: 3.7.5
- Operating System: Ubuntu 18.04

### Expected behavior

I ran two times the same training code using TemporalFusionTransformer, I expect a similar loss value on the validation dataset

### Actual behavior

However, result was different on each run.
I think it has to do with TemporalFusionTransformer because when running the same experiment with DeepAR model the problem is solved.

### Code to reproduce the problem

`pl.seed_everything(42)`

    warnings.simplefilter("error", category=SettingWithCopyWarning)

    data = generate_ar_data(seasonality=10.0, timesteps=400, n_series=100)
    data["static"] = "2"
    data["date"] = pd.Timestamp("2020-01-01") + pd.to_timedelta(data.time_idx, "D")
    validation = data.series.sample(20)

    max_encoder_length = 60
    max_prediction_length = 20

    training_cutoff = data["time_idx"].max() - max_prediction_length

    training = TimeSeriesDataSet(
        data[lambda x: ~x.series.isin(validation)],
        time_idx="time_idx",
        target="value",
        categorical_encoders={"series": NaNLabelEncoder().fit(data.series)},
        group_ids=["series"],
        static_categoricals=["static"],
        min_encoder_length=max_encoder_length,
        max_encoder_length=max_encoder_length,
        min_prediction_length=max_prediction_length,
        max_prediction_length=max_prediction_length,
        time_varying_unknown_reals=["value"],
        time_varying_known_reals=["time_idx"],
        target_normalizer=GroupNormalizer(groups=["series"]),
        add_relative_time_idx=False,
        add_target_scales=True,
        randomize_length=None,
    )

    validation = TimeSeriesDataSet.from_dataset(
        training,
        data[lambda x: x.series.isin(validation)],
        # predict=True,
        stop_randomization=True,
    )
    batch_size = 64
    train_dataloader = training.to_dataloader(train=True, batch_size=batch_size, num_workers=0)
    val_dataloader = validation.to_dataloader(train=False, batch_size=batch_size, num_workers=0)

    # save datasets
    training.save("training.pkl")
    validation.save("validation.pkl")

    early_stop_callback = EarlyStopping(monitor="val_loss", min_delta=1e-4, patience=5, verbose=False, mode="min")
    lr_logger = LearningRateMonitor()

    trainer = pl.Trainer(
        max_epochs=1,
        gpus=-1,
        weights_summary="top",
        gradient_clip_val=0.1,
        limit_train_batches=30,
        limit_val_batches=3,
        callbacks=[lr_logger, early_stop_callback],
    )

    model = TemporalFusionTransformer.from_dataset(
        training,
        learning_rate=0.1,
        hidden_size=32,
        dropout=0,
        loss=QuantileLoss([0.5]),
        output_size=1,
        log_interval=10,
        log_val_interval=3,
        # reduce_on_plateau_patience=3,
    )

    """model = DeepAR.from_dataset(
        training,
        learning_rate=0.1,
        hidden_size=32,
        dropout=0.1,
        loss=NormalDistributionLoss(),
        log_interval=10,
        log_val_interval=3,
        # reduce_on_plateau_patience=3,
    )"""
    print(f"Number of parameters in network: {model.size() / 1e3:.1f}k")

    trainer.fit(
        model,
        train_dataloader=train_dataloader,
        val_dataloaders=val_dataloader,
    )

    # calcualte mean absolute error on validation set
    actuals = torch.cat([y for x, y in iter(val_dataloader)])
    predictions = model.predict(val_dataloader)
    print(f"Mean absolute error of model: {(actuals - predictions).abs().mean()}")`

**Comment by jdb78** [comment]
Could you try in addition to the seed the trainer flag `deterministic=True` (see https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#reproducibility)?

**Comment by BorisAnc** [comment]
Hello,

Thanks for the suggestion. 
I've tried with the  the flag `deterministic=True` and also with no GPU.
This does not solve the issue with the TemporalFusionTransformer model.
